src/NUnitFramework/framework/Api/NUnitTestAssemblyRunner.cs

```diff
@@ -343,16 +343,22 @@ private void CreateTestExecutionContext(ITestListener listener)
 
             // Set the listener - overriding runners may replace this
             Context.Listener = listener;
-
+#if NETSTANDARD1_6
+            Context.Dispatcher = new MainThreadWorkItemDispatcher();
+#else
 #if PARALLEL
             int levelOfParallelism = GetLevelOfParallelism();
 
             if (levelOfParallelism > 0)
                 Context.Dispatcher = new ParallelWorkItemDispatcher(levelOfParallelism);
+            else if (Settings.ContainsKey(FrameworkPackageSettings.RunOnMainThread) &&
+                     (bool)Settings[FrameworkPackageSettings.RunOnMainThread])
+                Context.Dispatcher = new MainThreadWorkItemDispatcher();
```

Comment:
Ah - now I see why setting TestWorkers = 0 is required!

Given you could never run in parallel on the main thread, can I suggest that the `FrameworkPackageSettings.RunOnMainThread` check is done first, and that overrides any parallelism settings? That would hopefully simplify the logic - unless there's another reason to need `NumberOfTestWorkers` to be set to zero?

Reply:
I guess you are right. If the condition for `RunOnMainThread` comes first, I didn't see any reason for the `NumberOfTestWorkers` to be forced to 0, even though conceptually it implies that `NumberOfTestWorkers` is 0. So I am removing that part from FrameworkController.cs for now.